src/charm.py

```diff
@@ -615,15 +630,30 @@ def _smtp_connection_uri(self) -> str:
         else:
             return f"{method}://{username}:{password}@{server}:{port}/"
 
+    @property
+    def current_config_hash(self) -> Optional[int]:
+        return self._stored.config_hash
+
     @retry(
         wait=wait_exponential(multiplier=3, min=1, max=10),
         stop=stop_after_attempt(5),
         reraise=True,
         before=before_log(logger, logging.DEBUG),
     )
+    def _update_cm(self, content: str):
+        if self.unit.is_leader():
+            self.kratos_configmap.update({"kratos.yaml": content})
+
     def _update_config(self) -> None:
         conf = self._render_conf_file()
-        self._config_file_changed = self.kratos_configmap.update({"kratos.yaml": conf})
+        config_hash = utils.hash(conf)
+        if not config_hash != self.current_config_hash:
+            return
```

Comment:
feel free to ignore, but this seems clearer
```suggestion
        if config_hash == self.current_config_hash:
            return
```